### PR TITLE
fix(metadata): identify archives and folder ROMs by the right file

### LINKED ISSUE #3742: [Bug] Hasheous and Playmatch hash matching broken for multi-file archives since 4.9.0 (composite hash sent, but they index by largest file)
**RomM version**
4.9.0 and later (current master, 66596c8). 4.8.x is not affected.

**Describe the bug**
Some ROMs are zip files containing several files instead of one. Arcade (MAME/FBNeo) sets are the most common example, but this applies to any platform.

Since 4.9.0 (#3412), RomM hashes these archives by decompressing every file inside and computing one combined "composite" hash over all of them. That was Phase 1 of #3411, whose goal was populating the ROM-level hash fields for archives. Phase 2, deciding which hash each metadata provider should receive during lookup, was explicitly deferred ("this needs separate consideration" in the PR description) and never happened.

The unintended side effect: the metadata handlers for Hasheous and Playmatch read their hashes from the archive's file record, and that value silently changed from "hash of the largest file inside the archive" (pre-4.9.0) to the new composite hash. Both services index a multi-file archive by its largest internal file (the Playmatch developer confirmed the largest-file rule). The composite value matches nothing in their databases, so hash-based identification for multi-file archives silently fails on 4.9.0+. Those ROMs go unmatched or fall back to weaker filename matching.

Verified against the live APIs on 2026-08-07:

- **Hasheous**: the two MAME sets I tested return HTTP 404 on the composite and resolve on the largest member, to "Aero Fighters" and "Golden Axe: The Revenge of Death Adder" (signatures `MAMEArcade`, `FBNeo`).
- **Playmatch**: same rule for the datasets it carries, but it does not appear to index MAME sets at all. Both arcade sets return `NoMatch` on the composite *and* on the largest member, so **arcade is a Hasheous-only regression**. The Playmatch half reproduces on any non-arcade multi-file archive: a No-Intro ROM zipped with a readme returns `NoMatch` on the composite and `gameMatchType='SHA1'` on the largest member.

ScreenScraper still works in practice, but not because it matches the composite hash: its lookup endpoint falls back to the filename (`romnom`) when the hashes miss, and for arcade zips the filename is the set name, which ScreenScraper resolves directly.

Where this happens in the code (permalinks to master at 66596c8):

- Composite hash computed for archives: [`roms_handler.py#L450-L511`](https://github.com/rommapp/romm/blob/66596c85615967a4779e46023e37dab2cdc78053/backend/handler/filesystem/roms_handler.py#L450-L511)
- Sent to Hasheous: [`hasheous_handler.py#L258-L274`](https://github.com/rommapp/romm/blob/66596c85615967a4779e46023e37dab2cdc78053/backend/handler/metadata/hasheous_handler.py#L258-L274)
- Sent to Playmatch: [`playmatch_handler.py#L222-L232`](https://github.com/rommapp/romm/blob/66596c85615967a4779e46023e37dab2cdc78053/backend/handler/metadata/playmatch_handler.py#L222-L232)

**To Reproduce**

1. Run RomM 4.9.0+ with Hasheous and/or Playmatch enabled as metadata sources.
2. Add a multi-file zip to the library. Any ROM zipped together with a readme reproduces it on both services; a MAME arcade set reproduces it on Hasheous only, for the reason above.
3. Scan the platform.
4. The ROM gets no hash match, even though the same file matched on 4.8.x.

**Expected behavior**
Multi-file archives should be identified by Hasheous and Playmatch like they were before 4.9.0.

The fix should not require changing how ROMs are hashed. Since 4.9.0 the scanner already stores per-file hashes for every file inside an archive (`archive_members` on the file record), so the Hasheous and Playmatch handlers can select the largest member's hashes from there when the file is an archive. That completes Phase 2 of #3411 as originally planned, and works on existing libraries without re-hashing anything. The composite hash should stay as the stored ROM-level identity, which is what #3411 set out to add.

**Additional context**
Single-file zips are unaffected: the composite hash of one file equals that file's own hash, which is also what the pre-4.9.0 code produced. That is why most non-arcade libraries don't notice, and why this shows up primarily on arcade sets, where multi-file archives are the norm.

RetroAchievements matching is a separate pipeline and is unaffected by this change: for arcade, the RA hash is derived from the zip's filename, and for other platforms RAHasher cannot process multi-file zips at all regardless of what RomM computes.

Related: #3411 (Phase 1), #3412 (implementation).

### PR DESCRIPTION
**Pull request by Spinnich**
**Description**
<sup>Explain the changes or enhancements you are proposing with this pull request.</sup>

Fixes #3742.

Since 4.9.0 an archive's `RomFile` carries a **composite** digest streamed across every
entry in the archive (`a170649fe`, PR #3412). The metadata handlers were never updated
and kept reading their lookup digests straight off that record, so they silently
switched from "the ROM inside the zip" to "a hash of the whole archive". No ROM database
indexes that composite, so any zip with more than one member (arcade/MAME sets, but also
an ordinary ROM zipped next to a readme or patch) stopped matching by hash and fell back
to filename matching. Single-file zips were unaffected, which is why most libraries
never noticed.

This is a re-land of #4059, which implemented exactly this and was closed unmerged on
2026-08-02. The review request on it is still open and no formal review was ever
submitted, so I've kept half A deliberately close to that diff to keep the re-land
legible.

**What changed**

*Which hashes (the 4.9.0 regression).* New `RomFile.lookup_hashes` decides what a ROM
database should be queried with: `chd_sha1_hash` for CHDs, else the largest
`archive_members` entry, else the file's own digests. Hasheous and Playmatch both read
it. CHD behaviour for Hasheous is byte-identical to before; only the archive case moves.

*Which file (pre-existing, not a 4.9.0 regression).* Playmatch picked its file with
`next(file for file in files if file.file_size_bytes > 0)` — the *first* file with bytes,
not the largest, with no `is_top_level` filter. `iter_files` is a bare `os.walk` with no
sort and `Rom.files` has no `order_by`, so for a multi-disc folder ROM the file Playmatch
was asked about was arbitrary filesystem order and could differ between two machines
scanning the same library. It was also frequently the `.m3u` playlist, which carries no
game data and can only ever miss.

To be clear about scope: this half is **not** part of the 4.9.0 regression. 4.8.1 carries
the same three lines verbatim (`ca58028eb`, "playmatch should just use the first file",
2025-07-10). It's in here because fixing only the hash selection would leave multi-disc
folder ROMs still failing on Playmatch, and the two questions are the same question.
A new `_select_lookup_file` replaces both copies of that expression (lookup and
suggestion) with the largest top-level file, tie-broken on path.

**Worth looking at hardest**

1. `_select_lookup_file`'s `is_top_level` filter is a behaviour narrowing, not just a
   tie-break. A folder ROM whose files all live in subdirectories (`Game/Disc 1/game.bin`)
   now gets **no** Playmatch lookup at all, where before it got one against an arbitrary
   nested file. Hasheous and ScreenScraper already drop those libraries, so this aligns
   the three, but it is a real loss for that layout. Happy to add a last-resort fallback
   to the largest sizeable file if you'd rather keep the old reach.

2. `submit_manual_match_suggestion` now contributes the *member's* digests alongside the
   *archive's* `fileName`/`fileSize`. Harmless for lookups (a hash match outranks
   `FileNameAndSize`), but the contributed triple is internally inconsistent. Sending the
   member's name and size would be the fuller fix; I left it out of scope here and it's
   worth its own issue. Relatedly, the suggestion no longer falls back to
   `rom.md5_hash`/`rom.sha1_hash` when no file qualifies — it just returns. Contributing
   nothing to a public index beats contributing a composite that is a digest of no single
   file.

3. Playmatch now looks CHDs up by `chd_sha1_hash` instead of the container digests. This
   is right in principle and matches Hasheous, but Playmatch's CHD behaviour is the one
   leg I couldn't confirm with a live match.

**Deliberately unchanged**

ScreenScraper keeps sending the archive's own composite digests. Its `jeuInfos` lookup
falls back to `romnom` (the filename), which for an arcade zip is the set name, so those
still resolve; moving it would shift match results across many libraries. There's a new
test pinning that, so a future change there has to be a choice rather than a side effect.
RetroAchievements is a separate RAHasher pipeline and is untouched.

**No migration, no re-hash.** `archive_members` has been populated since 4.9.0 and
already carries the member digests, so this reads correct data off rows already in the
database.

**Correction to the issue body:** #3742 claims the arcade fixtures demonstrate the
Playmatch half. They don't — Playmatch does not index MAME sets at all and returns
`NoMatch` for them on *either* hash. An ordinary No-Intro ROM zipped with a readme is
what reproduces the Playmatch half; both are covered below.

---

> **AI assistance disclosure.** This PR was investigated, implemented, and written
> primarily by Claude Code — the root-cause analysis, the code, the tests, and this
> description. I directed the work, made the scope calls, reviewed every line, and ran
> the verification below myself.

**Checklist**
<sup>Please check all that apply.</sup>

- [x] I've tested the changes locally
- [x] I've updated relevant comments
- [x] I've assigned reviewers for this PR
- [x] I've added unit tests that cover the changes

<details>
<summary>Verification</summary>

**Live lookups against the real Hasheous and Playmatch APIs**, driving the real
`FSRomsHandler.get_rom_files()` (only `RAHasherService.calculate_hash` patched out) into
the real `lookup_rom`:

| Fixture | md5 sent before | md5 sent now | Hasheous | Playmatch |
| --- | --- | --- | --- | --- |
| `aerofgt.zip` (8 members) | `c5347ceb…` | `694853ae…` | **'Aero Fighters' hasheous 48064 / igdb 3147** (was HTTP 404) | `NoMatch` (expected) |
| `ga2.zip` (19 members) | `c6ed4e03…` | `0cd4f9af…` | **'Golden Axe: The Revenge of Death Adder' 51687 / igdb 5565** (was 404) | `NoMatch` (expected) |
| SMB3 zip (1 member, control) | `e883e0d4…` | identical | 'Super Mario Bros. 3' 25611 (unchanged) | `SHA1` igdb 1068 (unchanged) |
| SMB3 + readme (2 members) | `9f018a5b…` | `e883e0d4…` | **'Super Mario Bros. 3'** (was 404) | **`SHA1` igdb 1068** (was `NoMatch`) |

The two arcade `NoMatch`es are not a failure: Playmatch doesn't index MAME on either
hash. The SMB3+readme row is the fixture that demonstrates the Playmatch half.

Multi-disc folder ROM (`Metal Gear Solid (USA) (Rev 1)/`, two CHDs plus an `.m3u`):
picks `Disc 1`, not the playlist; sends the CHD's disc-data SHA-1, not the container's;
identical selection across 4 list orderings (order-dependent on master).

**Tests.** 15 new tests. All 7 in `tests/models/test_rom.py` failed on master with
`ImportError: cannot import name 'LookupHashes'`; the Hasheous payload test failed
sending `compositemd5/sha1/crc`; 6 of the 7 Playmatch tests failed on master (composite
hashes, `fileName='game.m3u'`, order-dependent payload, a nested file selected, a
composite POSTed as a suggestion). Edge cases covered: single-member archive unchanged,
`archive_members` `None` and `[]`, a member with a missing or `None` size, CHD
precedence, nested-file exclusion, and no-eligible-file.

Full backend suite: **2762 passed, 2 skipped**. `trunk fmt` + `trunk check --no-fix` on
all 7 files: no issues.

**No OpenAPI change** — proven, not assumed: dumped `app.openapi()` from a `master`
worktree and from this branch and diffed them. Identical, 26,574 lines each. So no
`npm run generate` and no frontend typecheck.

</details>

#### Screenshots (if applicable)

N/A — backend only. No UI, no i18n, no schema change. Evidence is the live-lookup table
above.

### CODE REVIEW

File: backend/models/rom.py [hunk 1]
```diff
@@ -165,6 +173,28 @@ def file_extension(self) -> str:
...
+
+        if self.archive_members:
+            largest = max(self.archive_members, key=lambda m: m.get("size") or 0)
+            return LookupHashes(
+                crc=largest.get("crc_hash"),
+                md5=largest.get("md5_hash"),
+                sha1=largest.get("sha1_hash"),
+            )
+
+        return LookupHashes(crc=self.crc_hash, md5=self.md5_hash, sha1=self.sha1_hash)
```
Comment (gantoine) [inline backend/models/rom.py:196]: very clean approach

**Review by gantoine — APPROVED** [review]
